File: backend/chain/licence_guard.py
"""
IRG Licence Guard — Django / Python runtime validator.

Perpetual licence, Ed25519-signed. Verification checks:
  1. Signature matches the embedded public key
  2. Deployment fingerprint matches what was signed
  3. Product is listed in permitted products
  4. Issuance timestamp is not in the future

Usage in Django:
  # In your AppConfig.ready():
  from .licence_guard import verify_licence_or_die
  verify_licence_or_die(product_code='GDP')

IPR Owner: Rohit Tidke | (c) 2026 Intech Research Group
"""

import base64
import hashlib
import json
import logging
import os
import socket
import threading
import time
import uuid
from pathlib import Path
from typing import Optional

logger = logging.getLogger("irg.licence")

# Hard-coded public key — replaced at build time by bootstrap.sh.
LICENCE_PUBLIC_KEY_HEX = os.environ.get(
    "IRG_LICENCE_PUBLIC_KEY_HEX",
    "0000000000000000000000000000000000000000000000000000000000000000",
)

LICENCE_TOKEN_PATH = os.environ.get("IRG_LICENCE_TOKEN_PATH", "/etc/irg/licence.token")
LICENCE_FINGERPRINT_SALT = b"irg-fingerprint-v1"
RECHECK_INTERVAL_SECONDS = 60 * 60


class LicenceError(Exception):
    pass


class _State:
    valid: bool = False
    reason: str = "not_checked"
    payload: Optional[dict] = None
    last_good_at: float = 0.0
    last_check_at: float = 0.0


STATE = _State()
_lock = threading.Lock()


def _primary_mac() -> str:
    try:
        return f"{uuid.getnode():012x}"
    except Exception:
        return "unknown"


def _chain_id() -> str:
    return os.environ.get("IRG_CHAIN_ID", "888101")


def _build_version() -> str:
    return os.environ.get("IRG_BUILD_VERSION", "v1.0")


def compute_deployment_fingerprint() -> str:
    h = hashlib.sha256()
    h.update(LICENCE_FINGERPRINT_SALT)
    h.update(_primary_mac().encode())
    h.update(b"|")
    h.update(socket.gethostname().encode())
    h.update(b"|")
    h.update(_chain_id().encode())
    h.update(b"|")
    h.update(_build_version().encode())
    return h.hexdigest()


def _b64url_decode(s: str) -> bytes:
    pad = "=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s + pad)


def _verify_ed25519(message: bytes, signature: bytes, pubkey: bytes) -> bool:
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
    except ImportError:
        logger.critical("cryptography library missing — cannot verify licence")
        return False
    try:
        pk = Ed25519PublicKey.from_public_bytes(pubkey)
        pk.verify(signature, message)
        return True
    except Exception:
        return False


def _parse_token(token: str) -> dict:
    parts = token.strip().split(".")
    if len(parts) != 2:
        raise LicenceError("malformed_token")
    payload_b = _b64url_decode(parts[0])
    sig = _b64url_decode(parts[1])
    try:
        pubkey = bytes.fromhex(LICENCE_PUBLIC_KEY_HEX)
    except ValueError:
        raise LicenceError("bad_public_key_config")
    if len(pubkey) != 32:
        raise LicenceError("bad_public_key_length")
    if not _verify_ed25519(payload_b, sig, pubkey):
        raise LicenceError("bad_signature")
    try:
        return json.loads(payload_b)
    except json.JSONDecodeError:
        raise LicenceError("bad_payload_json")


def _verify_once(product_code: str):
    path = Path(LICENCE_TOKEN_PATH)
    if not path.is_file():
        return False, f"token_not_found:{path}", None
    try:
        raw = path.read_text().strip()
        payload = _parse_token(raw)
    except (LicenceError, PermissionError) as e:
        return False, str(e), None

    if payload.get("v") != 2:
        return False, "unsupported_version", None

    now = int(time.time())
    if int(payload.get("iat", 0)) > now + 300:
        return False, "issued_in_future", payload

    if payload.get("fp", "").lower() != compute_deployment_fingerprint():
        return False, "fingerprint_mismatch", payload

    if product_code.upper() not in [p.upper() for p in payload.get("products", [])]:
        return False, "product_not_licensed", payload

    return True, "ok", payload


def verify_licence_or_die(product_code: str = "GDP") -> None:
    """Call from AppConfig.ready(). Terminates the process on invalid licence."""
    valid, reason, payload = _verify_once(product_code)
    with _lock:
        STATE.valid = valid
        STATE.reason = reason
        STATE.payload = payload
        STATE.last_check_at = time.time()
        if valid:
            STATE.last_good_at = time.time()

    if not valid:
        logger.critical(
            "[irg.licence] LICENCE INVALID (%s) — refusing to start. "
            "Contact the licensor.", reason,
        )
        if os.environ.get("IRG_LICENCE_TEST_MODE") == "1":
            raise LicenceError(reason)
        os._exit(2)

    logger.info(
        "[irg.licence] OK — %s licensed to %s (%s), serial %s",
        product_code,
        payload.get("name", "?"),
        payload.get("sub", "?"),
        payload.get("serial", "?"),
    )

    if not getattr(verify_licence_or_die, "_thread_started", False):
        verify_licence_or_die._thread_started = True
        threading.Thread(
            target=_recheck_loop, args=(product_code,),
            daemon=True, name="irg-licence-recheck",
        ).start()


def _recheck_loop(product_code: str) -> None:
    while True:
        time.sleep(RECHECK_INTERVAL_SECONDS)
        try:
            valid, reason, payload = _verify_once(product_code)
        except Exception as e:
            logger.warning("[irg.licence] recheck error: %s", e)
            continue
        with _lock:
            STATE.valid = valid
            STATE.reason = reason
            STATE.payload = payload
            STATE.last_check_at = time.time()
            if valid:
                STATE.last_good_at = time.time()
        if not valid:
            logger.error("[irg.licence] recheck failed: %s", reason)


def current_licence_info() -> dict:
    with _lock:
        return {
            "valid": STATE.valid,
            "reason": STATE.reason,
            "last_check_at": STATE.last_check_at,
            "last_good_at": STATE.last_good_at,
            "licensee": STATE.payload.get("name") if STATE.payload else None,
            "licensee_uid": STATE.payload.get("sub") if STATE.payload else None,
            "serial": STATE.payload.get("serial") if STATE.payload else None,
            "products": STATE.payload.get("products") if STATE.payload else [],
            "territory": STATE.payload.get("territory") if STATE.payload else [],
        }

File: functions/licence-guard.js
/**
 * IRG Licence Guard — plain Node.js for Firebase Cloud Functions.
 *
 * Perpetual licence, Ed25519-signed.
 *
 * Usage in functions/index.js:
 *   const { verifyLicenceOrDie, currentLicenceInfo } = require('./licence-guard');
 *   if (process.env.FUNCTIONS_EMULATOR !== 'true') {
 *     verifyLicenceOrDie('DAC');  // or 'GOV'
 *   }
 *   exports.licenceStatus = functions.https.onCall(async () => currentLicenceInfo());
 *
 * IPR Owner: Rohit Tidke | (c) 2026 Intech Research Group
 */

'use strict';

const crypto = require('crypto');
const fs = require('fs');
const os = require('os');

const LICENCE_PUBLIC_KEY_HEX =
  process.env.IRG_LICENCE_PUBLIC_KEY_HEX ||
  '0000000000000000000000000000000000000000000000000000000000000000';

const LICENCE_TOKEN_PATH =
  process.env.IRG_LICENCE_TOKEN_PATH || '/etc/irg/licence.token';
const FINGERPRINT_SALT = Buffer.from('irg-fingerprint-v1');
const RECHECK_INTERVAL_MS = 60 * 60 * 1000;

const STATE = {
  valid: false, reason: 'not_checked', payload: null,
  lastCheckAt: 0, lastGoodAt: 0,
};

function primaryMac() {
  const ifaces = os.networkInterfaces();
  for (const list of Object.values(ifaces)) {
    if (!list) continue;
    for (const addr of list) {
      if (!addr.internal && addr.mac && addr.mac !== '00:00:00:00:00:00') {
        return addr.mac.replace(/:/g, '').toLowerCase();
      }
    }
  }
  return 'unknown';
}

function chainId() { return process.env.IRG_CHAIN_ID || '888101'; }
function buildVersion() { return process.env.IRG_BUILD_VERSION || 'v1.0'; }

function computeDeploymentFingerprint() {
  const h = crypto.createHash('sha256');
  h.update(FINGERPRINT_SALT);
  h.update(primaryMac());
  h.update('|');
  h.update(os.hostname());
  h.update('|');
  h.update(chainId());
  h.update('|');
  h.update(buildVersion());
  return h.digest('hex');
}

function b64urlDecode(s) {
  const pad = s.length % 4 === 0 ? '' : '='.repeat(4 - (s.length % 4));
  return Buffer.from(s + pad, 'base64url');
}

function ed25519PubKeyFromHex(hex) {
  const raw = Buffer.from(hex, 'hex');
  if (raw.length !== 32) throw new Error('bad_public_key_length');
  const der = Buffer.concat([Buffer.from('302a300506032b6570032100', 'hex'), raw]);
  return crypto.createPublicKey({ key: der, format: 'der', type: 'spki' });
}

function verifySignature(message, signature) {
  try {
    return crypto.verify(null, message, ed25519PubKeyFromHex(LICENCE_PUBLIC_KEY_HEX), signature);
  } catch {
    return false;
  }
}

function parseToken(token) {
  const parts = token.trim().split('.');
  if (parts.length !== 2) throw new Error('malformed_token');
  const payload = b64urlDecode(parts[0]);
  const sig = b64urlDecode(parts[1]);
  if (!verifySignature(payload, sig)) throw new Error('bad_signature');
  return JSON.parse(payload.toString('utf8'));
}

function verifyOnce(productCode) {
  let raw;
  try {
    raw = fs.readFileSync(LICENCE_TOKEN_PATH, 'utf8').trim();
  } catch {
    return { valid: false, reason: `token_not_found:${LICENCE_TOKEN_PATH}`, payload: null };
  }
  let payload;
  try {
    payload = parseToken(raw);
  } catch (e) {
    return { valid: false, reason: e.message || 'parse_error', payload: null };
  }
  if (payload.v !== 2) return { valid: false, reason: 'unsupported_version', payload };

  const now = Math.floor(Date.now() / 1000);
  if (payload.iat > now + 300) return { valid: false, reason: 'issued_in_future', payload };

  if ((payload.fp || '').toLowerCase() !== computeDeploymentFingerprint()) {
    return { valid: false, reason: 'fingerprint_mismatch', payload };
  }

  const products = (payload.products || []).map((p) => String(p).toUpperCase());
  if (!products.includes(String(productCode).toUpperCase())) {
    return { valid: false, reason: 'product_not_licensed', payload };
  }

  return { valid: true, reason: 'ok', payload };
}

let recheckStarted = false;

function verifyLicenceOrDie(productCode) {
  const r = verifyOnce(productCode || 'DAC');
  STATE.valid = r.valid;
  STATE.reason = r.reason;
  STATE.payload = r.payload;
  STATE.lastCheckAt = Date.now();
  if (r.valid) STATE.lastGoodAt = Date.now();

  if (!r.valid) {
    console.error(
      `[irg.licence] LICENCE INVALID (${r.reason}) — refusing to start. ` +
      `Contact the licensor.`,
    );
    if (process.env.IRG_LICENCE_TEST_MODE === '1') {
      throw new Error(r.reason);
    }
    process.exit(2);
  }

  console.info(
    `[irg.licence] OK — ${productCode} licensed to ${r.payload.name} ` +
    `(${r.payload.sub}), serial ${r.payload.serial}`,
  );

  if (!recheckStarted) {
    recheckStarted = true;
    const t = setInterval(() => {
      const rr = verifyOnce(productCode || 'DAC');
      STATE.valid = rr.valid;
      STATE.reason = rr.reason;
      STATE.payload = rr.payload;
      STATE.lastCheckAt = Date.now();
      if (rr.valid) STATE.lastGoodAt = Date.now();
      if (!rr.valid) console.error(`[irg.licence] recheck failed: ${rr.reason}`);
    }, RECHECK_INTERVAL_MS);
    if (typeof t.unref === 'function') t.unref();
  }
}

function currentLicenceInfo() {
  return {
    valid: STATE.valid,
    reason: STATE.reason,
    lastCheckAt: STATE.lastCheckAt,
    lastGoodAt: STATE.lastGoodAt,
    licensee: STATE.payload ? STATE.payload.name : null,
    licenseeUid: STATE.payload ? STATE.payload.sub : null,
    serial: STATE.payload ? STATE.payload.serial : null,
    products: STATE.payload ? STATE.payload.products : [],
    territory: STATE.payload ? STATE.payload.territory : [],
  };
}

module.exports = {
  verifyLicenceOrDie, currentLicenceInfo, computeDeploymentFingerprint,
};
